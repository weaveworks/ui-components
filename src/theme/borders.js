import { kebabCase, forEach } from 'lodash';

export const borderRadius = {
  circle: '50%',
  none: '0',
  soft: '2px',
};

// Collects all theme border radii as SCSS vars.
export function themeBorderRadiiAsScss() {
  const themeBorderRadii = [];

  forEach(borderRadius, (value, name) => {
    themeBorderRadii.push(`$border-radius-${kebabCase(name)}: ${value}`);
  });

  return themeBorderRadii;
}
