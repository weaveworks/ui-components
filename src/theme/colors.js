import { isString, kebabCase, forEach } from 'lodash';

export const colors = {
  // Neutral Colors
  black: 'hsl(0, 0%, 10%)', // #1a1a1a
  blue50: 'hsl(191, 100%, 97%)', // #f0fcff
  blue200: 'hsl(191, 100%, 80%)', // #99ecff
  blue400: 'hsl(191, 100%, 50%)', // #00d2ff
  blue600: 'hsl(191, 100%, 40%)', // #00a7cc
  blue700: 'hsl(191, 100%, 35%)', // #0092b3
  blue800: 'hsl(191, 100%, 30%)', // #007d99
  graphThemes: {
    blue: [
      '#c7e9b4',
      '#7ecdbb',
      '#1eb5eb',
      '#1d91bf',
      '#235fa9',
      '#253393',
      '#084181',
    ],
    mixed: [
      '#c7e9b4',
      '#c1d4e7',
      '#7ecdbb',
      '#9fbddb',
      '#1eb5eb',
      '#8d95c6',
      '#1d91bf',
      '#8282ab',
      '#235fa9',
      '#89429e',
      '#253393',
      '#800f7a',
      '#084181',
      '#0b0533',
    ],
    purple: [
      '#c1d4e7',
      '#9fbddb',
      '#8d95c6',
      '#8282ab',
      '#89429e',
      '#800f7a',
      '#0b0533',
    ],
  }, // Used by PrometheusGraph component
  gray50: 'hsl(0, 0%, 96%)', // #f4f4f4
  gray100: 'hsl(0, 0%, 90%)', // #e6e6e6
  gray200: 'hsl(0, 0%, 80%)', // #cccccc

  // #737373
  gray600: 'hsl(0, 0%, 45%)',
  green500: 'hsl(161, 54%, 48%)', // #38bd93
  orange500: 'hsl(13, 100%, 50%)', // #ff3700
  orange600: 'hsl(13, 100%, 40%)', // #cc2c00
  orange700: 'hsl(13, 100%, 35%)', // #b32700
  // Accent Colors
  // #992100
  orange800: 'hsl(13, 100%, 30%)',
  promQL: {
    attrName: '#00a4db',
    /**
     * GHColors theme by Avi Aryan (http://aviaryan.in)
     * Inspired by Github syntax coloring
     */
    comment: '#bbbbbb',
    deleted: '#9a050f',
    entity: '#36acaa',
    function: '#dc322f',
    metricName: '#2aa198',
    punctuation: '#393a34',
    // Dropdown colors,
    salmon: '#ff7c7c',
    string: '#e3116c',
    tag: '#00009f',
  }, // PromQL
  purple25: 'hsl(240, 20%, 98%)', // #fafafc
  purple50: 'hsl(240, 20%, 95%)', // #eeeef4
  purple100: 'hsl(240, 20%, 90%)', // #dfdfea
  purple200: 'hsl(240, 20%, 75%)', // #b1b1cb
  purple300: 'hsl(240, 20%, 65%)', // #9494b8
  purple400: 'hsl(240, 20%, 60%)', // #8585ad

  purple500: 'hsl(240, 20%, 50%)', // #666699
  purple600: 'hsl(240, 20%, 45%)', // #5b5b88
  purple700: 'hsl(240, 20%, 35%)', // #47476b
  purple800: 'hsl(240, 20%, 30%)', // #3d3d5c
  // Primary Colors
  // #32324b
  purple900: 'hsl(240, 20%, 25%)',

  // Third-party specific colors - not to be used in the theme!
  thirdParty: {
    azure: '#3769bb',
    // Google single-click login
    cornflowerBlue: '#4285f4',
  },
  // #ffffff
  white: 'hsl(0, 0%, 100%)',
  // #d4ab27
  yellow500: 'hsl(46, 69%, 49%)',
};

// Flattens and collects all theme colors as SCSS vars
export function themeColorsAsScss() {
  const ignoreKeys = ['graphThemes'];
  const themeColors = [];

  forEach(colors, (value, name) => {
    const colorPrefix = `$color-${kebabCase(name)}`;
    if (ignoreKeys.includes(name)) return;

    if (isString(value)) {
      themeColors.push(`${colorPrefix}: ${value}`);
    } else {
      forEach(value, (innerValue, subname) => {
        if (ignoreKeys.includes(subname)) return;
        themeColors.push(`${colorPrefix}-${kebabCase(subname)}: ${innerValue}`);
      });
    }
  });

  return themeColors;
}
