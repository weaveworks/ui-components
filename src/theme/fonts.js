import { kebabCase, forEach } from 'lodash';

export const fontFamilies = {
  monospace: "'Roboto Mono', monospace",
  regular: "'proxima-nova', Helvetica, Arial, sans-serif",
};

export const fontSizes = {
  extraLarge: '32px',
  huge: '48px',
  large: '22px',
  normal: '16px',
  small: '14px',
  tiny: '12px',
};

// Collects all theme font vars as SCSS vars.
export function themeFontsAsScss() {
  const themeFonts = [];

  forEach(fontFamilies, (value, name) => {
    themeFonts.push(`$font-family-${kebabCase(name)}: ${value}`);
  });

  forEach(fontSizes, (value, name) => {
    themeFonts.push(`$font-size-${kebabCase(name)}: ${value}`);
  });

  return themeFonts;
}
