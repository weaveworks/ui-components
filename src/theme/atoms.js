import { colors } from './colors';

// component-specific
export const atoms = {
  Alert: {
    error: {
      background: colors.orange600,
      color: colors.white,
    },
    info: {
      background: colors.blue600,
      color: colors.white,
    },
    success: {
      background: colors.green500,
      color: colors.white,
    },
    warning: {
      background: colors.yellow500,
      color: colors.white,
    },
  },
  Button: {
    danger: {
      background: colors.orange600,
      color: colors.white,
      hoverBackground: colors.orange700,
      hoverColor: colors.white,
    },
    default: {
      background: colors.white,
      color: colors.black,
      hoverBackground: colors.gray50,
      hoverColor: colors.purple800,
    },
    disabled: {
      background: colors.gray50,
      color: colors.gray600,
      hoverBackground: colors.gray50,
      hoverColor: colors.gray600,
    },
    primary: {
      background: colors.blue700,
      color: colors.white,
      hoverBackground: colors.blue800,
      hoverColor: colors.white,
    },
  },
};
