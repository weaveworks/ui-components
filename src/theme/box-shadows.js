import { colors } from './colors';

export const boxShadow = {
  heavy: '0 10px 20px rgba(0,0,0,0.19), 0 6px 6px rgba(0,0,0,0.23)',
  light: '0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24)',
  none: 'none',
  selected: `0px 0px 2px 2px ${colors.purple400}`,
};
