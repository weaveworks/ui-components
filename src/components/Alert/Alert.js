import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import { fromAtoms } from '../../utils/theme';

const StyledAlert = styled.div`
  background: ${fromAtoms('Alert', 'type', 'background')};
  border-radius: ${props => props.theme.borderRadius.soft};
  border: 2px solid ${fromAtoms('Alert', 'type', 'borderColor')};
  color: ${fromAtoms('Alert', 'type', 'color')};
  margin: 8px 0 8px 0;
  min-height: 1em;
  padding: 1em;
  transition: opacity 0.2s linear;

  a {
    color: ${fromAtoms('Alert', 'type', 'color')};
    text-decoration: underline;

    &:hover {
      text-decoration: none;
    }
  }
`;

const CloseIcon = styled.i`
  float: right;
  cursor: pointer;
  margin-left: 1em;
`;

/**
 * An alert to let the user know about an action or state.
 * ```javascript
 * import { Alert } from 'weaveworks-ui-components';
 *
 * export default function() {
 *  return (
 *    <div>
 *     <Alert type="warning" visible={true}>
 *       Warning: Hmm, this is not good, but it is not terrible.
 *     </Alert>
 *    </div>
 *   )
 * }
 * ```
 */
function Alert(props) {
  const { onClose, children, visible } = props;
  if (!visible) {
    return null;
  }

  return (
    <StyledAlert {...props}>
      {children}
      {onClose && (
        <CloseIcon
          visible={visible}
          onClick={onClose}
          className="fa fa-remove"
        />
      )}
    </StyledAlert>
  );
}

Alert.propTypes = {
  /**
   * Set the color scheme to indicate the nature of the alert.
   */
  type: PropTypes.oneOf(['default', 'info', 'success', 'warning', 'error']),

  /**
   * Toggle whether the alert is shown
   */
  visible: PropTypes.bool,
  /**
   * Callback that runs with the alert is dismissed by clicking the 'X'
   */
  onClose: PropTypes.func,
};

Alert.defaultProps = {
  type: 'default',
  visible: true,
};

export default Alert;
