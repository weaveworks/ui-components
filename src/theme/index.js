import { colors, themeColorsAsScss } from './colors';
import { fontSizes, fontFamilies, themeFontsAsScss } from './fonts';
import { borderRadius, themeBorderRadiiAsScss } from './borders';
import { layers, themeLayersAsScss } from './layers';
import { spacing, themeSpacingsAsScss } from './spacings';
import { overlayIconSize, textColor, themeMiscVarsAsScss } from './misc';
import { boxShadow } from './box-shadows';
import { atoms } from './atoms';

export function themeVarsAsScss() {
  const themeVariables = []
    .concat(themeColorsAsScss())
    .concat(themeLayersAsScss())
    .concat(themeFontsAsScss())
    .concat(themeBorderRadiiAsScss())
    .concat(themeSpacingsAsScss())
    .concat(themeMiscVarsAsScss());
  return `${themeVariables.join('; ')};`;
}

const theme = {
  // Component-specific
  atoms,

  // Borders
  borderRadius,

  // Box shadows
  boxShadow,

  // Colors
  colors,

  // Fonts
  fontFamilies,
  fontSizes,

  // z-index layers
  layers,

  // Misc
  overlayIconSize,

  // Spacings
  spacing,
  textColor,
};

export default theme;
