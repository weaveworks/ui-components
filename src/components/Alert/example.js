import React from 'react';

import { Example } from '../../utils/example';
import Button from '../Button';

import Alert from '.';

const initial = {
  default: true,
  info: true,
  success: true,
  warning: true,
  error: true,
};

export default class AlertExample extends React.Component {
  constructor() {
    super();
    this.state = initial;
  }
  render() {
    const onClose = id => () => this.setState({ [id]: false });
    return (
      <div>
        <Example>
          <Alert visible={this.state.default} onClose={onClose('default')}>
            Default: This is a default alert. It is just giving you some info.
          </Alert>
        </Example>
        <Example>
          <Alert
            type="info"
            visible={this.state.info}
            onClose={onClose('info')}
          >
            Info: Please pay attention to this.
          </Alert>
        </Example>
        <Example>
          <Alert
            type="success"
            visible={this.state.success}
            onClose={onClose('success')}
          >
            Success: Great job! Wow! Much success!
          </Alert>
        </Example>
        <Example>
          <Alert
            type="warning"
            visible={this.state.warning}
            onClose={onClose('warning')}
          >
            Warning: Hmm, this is not good, but it is not terrible.
          </Alert>
        </Example>
        <Example>
          <Alert
            type="error"
            visible={this.state.error}
            onClose={onClose('error')}
          >
            Error: Wow you really screwed this up...
          </Alert>
        </Example>
        <Button onClick={() => this.setState(initial)} text="Reset" />
      </div>
    );
  }
}
