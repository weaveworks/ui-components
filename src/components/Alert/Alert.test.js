import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { withTheme } from '../../utils/theme';

import Alert from './Alert';

const c = renderer.create;

describe('<Alert />', () => {
  describe('snapshots', () => {
    it('should not render if visible="false"', () => {
      const tree = c(
        withTheme(<Alert visible={false}>Success</Alert>)
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders success', () => {
      const tree = c(withTheme(<Alert type="success">Success</Alert>)).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders warning', () => {
      const tree = c(withTheme(<Alert type="warning">Warning</Alert>)).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders error', () => {
      const tree = c(withTheme(<Alert type="error">Error</Alert>)).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders default', () => {
      const tree = c(withTheme(<Alert>Default</Alert>)).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  it('closes when the "x" is clicked', () => {
    const spy = jest.fn();
    const alert = mount(withTheme(<Alert onClose={spy}>My alert</Alert>));
    alert
      .find('.fa-remove')
      .first()
      .simulate('click');
    expect(spy.mock.calls.length).toEqual(1);
  });
});
