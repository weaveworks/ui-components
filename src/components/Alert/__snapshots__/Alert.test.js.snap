// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> snapshots renders default 1`] = `
.c0 {
  background: #f8f8f8;
  border-radius: 4px;
  border: 2px solid #d2d2d2;
  color: #1a1a1a;
  margin: 8px 0 8px 0;
  min-height: 1em;
  padding: 1em;
  -webkit-transition: opacity 0.2s linear;
  transition: opacity 0.2s linear;
}

.c0 a {
  color: #1a1a1a;
  text-decoration: underline;
}

.c0 a:hover {
  text-decoration: none;
}

<div
  className="c0"
  type="default"
>
  Default
</div>
`;

exports[`<Alert /> snapshots renders error 1`] = `
.c0 {
  background: #c7254e;
  border-radius: 4px;
  border: 2px solid #861935;
  color: #fff;
  margin: 8px 0 8px 0;
  min-height: 1em;
  padding: 1em;
  -webkit-transition: opacity 0.2s linear;
  transition: opacity 0.2s linear;
}

.c0 a {
  color: #fff;
  text-decoration: underline;
}

.c0 a:hover {
  text-decoration: none;
}

<div
  className="c0"
  type="error"
>
  Error
</div>
`;

exports[`<Alert /> snapshots renders success 1`] = `
.c0 {
  background: #38bd93;
  border-radius: 4px;
  border: 2px solid #278265;
  color: #fff;
  margin: 8px 0 8px 0;
  min-height: 1em;
  padding: 1em;
  -webkit-transition: opacity 0.2s linear;
  transition: opacity 0.2s linear;
}

.c0 a {
  color: #fff;
  text-decoration: underline;
}

.c0 a:hover {
  text-decoration: none;
}

<div
  className="c0"
  type="success"
>
  Success
</div>
`;

exports[`<Alert /> snapshots renders warning 1`] = `
.c0 {
  background: #d4ab27;
  border-radius: 4px;
  border: 2px solid #93771b;
  color: #fff;
  margin: 8px 0 8px 0;
  min-height: 1em;
  padding: 1em;
  -webkit-transition: opacity 0.2s linear;
  transition: opacity 0.2s linear;
}

.c0 a {
  color: #fff;
  text-decoration: underline;
}

.c0 a:hover {
  text-decoration: none;
}

<div
  className="c0"
  type="warning"
>
  Warning
</div>
`;

exports[`<Alert /> snapshots should not render if visible="false" 1`] = `null`;
